system-addon/common/Reducers.jsm

```diff
@@ -91,18 +91,26 @@ function TopSites(prevState = INITIAL_STATE.TopSites, action) {
       }
       return Object.assign({}, prevState, {initialized: true, rows: action.data});
     case at.TOP_SITES_EDIT:
-      return Object.assign({}, prevState, {editForm: {index: action.data.index}});
+      return Object.assign({}, prevState, {editForm: {index: action.data.index}, screenshotRequestFailed: false, screenshotPreview: null});
```

Comment:
> the url being requested to be associated with the preview somehow
If you add a new topsite you can do a "preview request" without adding in the URL first.
Instead of having two different ways to handle for this I had this capture all scenario.
Should we still try and fix it?

Reply:
In another comment https://github.com/mozilla/activity-stream/pull/3891#discussion_r170874864, you mentioned issues around needing to know whether the user has reverted back to the original custom image after a failed screenshot. If we tracked the preview url request / response failure and had the original site object, we would be able to determine that the failed screenshot is for a different preview url and that the current typed in preview url is the original value.

Reply:
> Should the screenshot stuff not be part of editForm? 

Yes

> I was referring to the preview url being requested not the url of the site

Sanity check: when replying with the screenshot preview data we should send the preview URL as well to check that we are replying to the correct request

I have replied to something completely different previously.

Reply:
> Yes

What's the reasoning for not having screenshot/preview stuff inside `editForm` state?

Reply:
> Sanity check: when replying with the screenshot preview data we should send the preview URL as well to check that we are replying to the correct request?

Right. I'm not sure if we need to save both the request and response as theoretically we could just ignore responses that aren't for the current request. E.g., when making a request, the page's redux state updates with the requested preview url and also sends to main. If there's a second request to main, the page's state again updates, and now there's 2 outstanding preview requests. I don't think backgroundpagethumbs exposes a way to cancel easily, so let's just let both requests finish for now, and the response has the thumbnail data and the requested url, but only the response that we now care about (the second url, which might have responded sooner).